state/accumulator_test.go

```diff
@@ -630,6 +655,73 @@ func TestCalculateNewSnapshotDupe(t *testing.T) {
 	}
 }
 
+// Test that you can accumulate the same room with the same partial sequence of timeline events and
+// state is updated correctly. This relies on postgres blocking subsequent transactions sensibly.
+func TestAccumulatorConcurrency(t *testing.T) {
+	roomID := "!TestAccumulatorConcurrency:localhost"
+	roomEvents := []json.RawMessage{
+		[]byte(`{"event_id":"AA", "type":"m.room.create", "state_key":"", "content":{"creator":"@me:localhost"}}`),
+		[]byte(`{"event_id":"BB", "type":"m.room.member", "state_key":"@me:localhost", "content":{"membership":"join"}}`),
+		[]byte(`{"event_id":"CC", "type":"m.room.join_rules", "state_key":"", "content":{"join_rule":"public"}}`),
+	}
+	db, close := connectToDB(t)
+	defer close()
+	accumulator := NewAccumulator(db)
+	_, err := accumulator.Initialise(roomID, roomEvents)
+	if err != nil {
+		t.Fatalf("failed to Initialise accumulator: %s", err)
+	}
+
+	// spin off N goroutines to insert 1,2,3,4,5 in varying batches which can appear in the wild e.g [1,2] [1,2,3,4]
```

Comment:
The batches we use have the property that for any two batches, one is a prefix of the other (e.g. [1,2] is a prefix of [1,2,3,4]).

Is it possible that we could race to accumulate batches without this property, e.g. batches [1,2] and [3,5]? Does that cause problems if so?

(I think we have talked about it, but it would be good to write something down here as I keep forgetting.)